client/src/unifyfs-fixed.c

```diff
@@ -543,78 +621,77 @@ static int unifyfs_logio_chunk_write(
+     */
+    if (unifyfs_flatten_writes) {
+        seg_tree_add(&meta->seg_tree, pos, pos + count - 1, log_offset);
     }
 
-    /* add new index entries if needed */
-    while (cur_idx.length > 0) {
-        /* remaining entries we can fit in the shared memory region */
-        off_t remaining_entries = unifyfs_max_index_entries - num_entries;
+    /* Update our write metadata with the new write */
+    return (unifyfs_logio_add_write_meta_to_index(gfid, pos, log_offset,
+        count));
+}
 
-        /* if we have filled the key/value buffer, flush it to server */
-        if (0 == remaining_entries) {
-            /* index buffer is full, flush it */
-            int ret = invoke_client_fsync_rpc(cur_idx.gfid);
-            if (ret != UNIFYFS_SUCCESS) {
-                /* something went wrong when trying to flush key/values */
-                LOGERR("failed to flush key/value index to server");
-                return UNIFYFS_ERROR_IO;
-            }
+/*
+ * Clear all entries in the log index.  This only clears the metadata,
+ * not the data itself.
+ */
+static void
+unifyfs_clear_index(void)
+{
+    size_t index_size;
+    index_size = sizeof(*unifyfs_indices.index_entry) *
+        (*unifyfs_indices.ptr_num_entries);
 
-            /* flushed, clear buffer and refresh number of entries
-             * and number remaining */
-            num_entries = 0;
-            *(unifyfs_indices.ptr_num_entries) = num_entries;
-            remaining_entries = unifyfs_max_index_entries - num_entries;
-        }
+    memset(unifyfs_indices.index_entry, 0, index_size);
+    *unifyfs_indices.ptr_num_entries = 0;
+}
 
-        /* split any remaining write index at boundaries of
-         * unifyfs_key_slice_range */
-        off_t used_entries = 0;
-        int split_rc = unifyfs_split_index(&cur_idx,
-            unifyfs_key_slice_range, &idxs[num_entries],
-            remaining_entries, &used_entries);
-        if (split_rc != UNIFYFS_SUCCESS) {
-            /* in this case, we have copied data to the log,
-             * but we failed to generate index entries,
-             * we're returning with an error and leaving the data
-             * in the log */
-            LOGERR("failed to split write index");
-            return UNIFYFS_ERROR_IO;
+/*
+ * Remove all entries in the current index and re-write it using the write
+ * metadata stored in all the file's seg_trees.  This only re-writes the
+ * metadata in the index.  All the actual data is still kept in the log and
+ * will be referenced correctly by the new metadata.
+ *
+ * After this function is done 'unifyfs_indices' will have been totally
+ * re-written.  The writes in the index will be flattened, non-overlapping,
+ * and sequential, for each file, one after another.  All seg_trees will be
+ * cleared.
+ *
+ * This function is typically called on an fsync().
+ */
+void unifyfs_rewrite_index_from_seg_tree(void)
+{
+    struct seg_tree_node* node;
+    int fid, gfid;
```

Comment:
style: let's move these three declarations down to the lines where the variables are used:
```
int fid = unifyfs_fds[i].fid;
```